bench/micro/Bench/Database/LSMTree/Internal/IndexCompact.hs

```diff
@@ -89,3 +94,74 @@ constructIndexCompact (ChunkSize csize) (RFPrecision rfprec, apps) = runST $ do
     mapM_ (`append` ica) apps
     (_, index) <- unsafeEnd ica
     pure index
+
+{-------------------------------------------------------------------------------
+  Benchmarks for UTxO keys that are /almost/ uniformly distributed
+-------------------------------------------------------------------------------}
+
+-- | UTXO keys are not truly uniformly distrbuted. The 'txId' is a uniformly
+-- distributed hash, but the same 'txId' can appear in multiple UTXO keys, but
+-- with a different 'txIx'. In the worst case, this means that we have a clash
+-- in the compact index for /every page/. The following benchmarks show
+benchNonUniform :: Benchmark
+benchNonUniform =
+    bgroup "non-uniformity" [
+        -- construction
+        env (pure $ (0, appsWithNearDups (mkStdGen 17) 1000)) $ \as ->
+          bench ("construct appsWithNearDups") $ whnf (constructIndexCompact 1000) as
+      , env (pure $ (0, appsWithoutNearDups (mkStdGen 17) 1000)) $ \as ->
+          bench ("construct appsWithoutNearDups") $ whnf (constructIndexCompact 1000) as
+        -- search
+      , env ( let ic = constructIndexCompact 100 (0, appsWithNearDups (mkStdGen 17) 1000)
+                  g  = mkStdGen 42
+                  ks = serialiseKey <$> uniformWithReplacement @UTxOKey g 1000
+              in  pure (ic, ks) ) $ \ ~(ic, ks) ->
+          bench "search appsWithNearDups" $ whnf (searches ic) ks
+      , env ( let ic = constructIndexCompact 100 (0, appsWithoutNearDups (mkStdGen 17) 1000)
+                  g  = mkStdGen 42
+                  ks = serialiseKey <$> uniformWithReplacement @UTxOKey g 1000
+              in  pure (ic, ks) ) $ \ ~(ic, ks) ->
+          bench "search appsWithoutNearDups" $ whnf (searches ic) ks
+      ]
+
+-- | 'Append's with truly uniformly distributed UTXO keys.
+appsWithoutNearDups ::
+     StdGen
+  -> Int -- ^ Number of pages
+  -> [Append]
+appsWithoutNearDups g n = assert (suggestRangeFinderPrecision n == 0) $
+    let ks  = uniformWithoutReplacement @UTxOKey g n
+        ks' = List.sort ks
+        -- append a dummy UTXO key because appsWithNearDups does so too.
+        ps  = groupsOfN 2 (UTxOKey 0 0 : ks')
+    in  fmap fromNE ps
+
+-- | 'Append's with worst-case near-duplicates. Each page boundary splits UTXO
+-- keys with the same 'txId' but different a 'txIx'.
+appsWithNearDups ::
+     StdGen
+  -> Int -- ^ Number of pages
+  -> [Append]
+appsWithNearDups g n = assert (suggestRangeFinderPrecision n == 0) $
+    let ks  = uniformWithoutReplacement @UTxOKey g (n `div` 2)
+        ks' = flip concatMap (List.sort ks) $ \k -> [k {txIx = 0}, k {txIx = 1}]
+        -- append a dummy UTXO key so that each pair of near-duplicate keys is
+        -- split between pages. That is, the left element of the pair is the
+        -- maximum key in a page, and the right element of the pair is the
+        -- minimum key on the next page.
+        ps  = groupsOfN 2 (UTxOKey 0 0 : ks')
+    in  fmap fromNE ps
```

Comment:
Ah, that `` `div` 2`` is crucial. I missed it at first and was looking at this a for a bit to be sure the number of pages are the same. Maybe an `assert (length ps == n)` in both functions could make this clearer.

Reply:
Updated

Reply:
I changed it to `*2`, that makes more sense. I didn't have this in earlier versions, so the benchmarks didn't have the same size inputs. The `assert` you suggested also helped me catch a bug. I've updated the numbers in the description to reflect the changes

Reply:
Ah, nice!